universal-application-tool-0.0.1/test/services/applicant/ReadOnlyApplicantProgramServiceImplTest.java

```diff
@@ -8,60 +8,89 @@
-
-  ProgramDefinition programDefinition =
-      ProgramDefinition.builder()
-          .setId(123L)
-          .setName("The Program")
-          .setDescription("This program is for testing.")
-          .addBlockDefinition(blockA)
-          .addBlockDefinition(blockB)
-          .build();
-
-  ReadOnlyApplicantProgramService subject;
+  private QuestionDefinition nameQuestion;
+  private QuestionDefinition colorQuestion;
+  private QuestionDefinition addressQuestion;
+  private ApplicantData applicantData;
+  private ProgramDefinition programDefinition;
+  private ReadOnlyApplicantProgramService subject;
 
   @Before
   public void setUp() {
-    subject =
-        new ReadOnlyApplicantProgramServiceImpl(
-            new Applicant().getApplicantData(), programDefinition);
+    applicantData = new ApplicantData();
+    nameQuestion = TestQuestionBank.applicantName().getQuestionDefinition();
+    colorQuestion = TestQuestionBank.applicantFavoriteColor().getQuestionDefinition();
+    addressQuestion = TestQuestionBank.applicantAddress().getQuestionDefinition();
+    programDefinition =
+        ProgramBuilder.newProgram()
+            .withBlock("Block one")
+            .withQuestionDefinition(nameQuestion)
+            .withBlock("Block two")
+            .withQuestionDefinition(colorQuestion)
+            .withQuestionDefinition(addressQuestion)
+            .buildDefinition();
+    subject = new ReadOnlyApplicantProgramServiceImpl(applicantData, programDefinition);
   }
 
   @Test
   public void getCurrentBlockList_getsTheApplicantSpecificBlocksForTheProgram() {
     ImmutableList<Block> blockList = subject.getCurrentBlockList();
 
-    assertThat(blockList.size()).isEqualTo(2);
+    assertThat(blockList).hasSize(2);
     Block block = blockList.get(0);
-    assertThat(block.getId()).isEqualTo(123L);
-    assertThat(block.getName()).isEqualTo("Block Name");
-    assertThat(block.getDescription()).isEqualTo("Block Description");
+    assertThat(block.getName()).isEqualTo("Block one");
+  }
+
+  @Test
+  public void getCurrentBlockList_doesNotIncludeCompleteBlocks() {
+    // Answer block one questions
+    answerNameQuestion();
+
+    ImmutableList<Block> blockList = subject.getCurrentBlockList();
+
+    assertThat(blockList).hasSize(1);
+    assertThat(blockList.get(0).getName()).isEqualTo("Block two");
+  }
+
+  @Test
+  public void getCurrentBlockList_returnsEmptyListIfAllBlocksComplete() {
+    // Answer all questions
+    answerNameQuestion();
+    answerColorQuestion();
+    answerAddressQuestion();
+
+    ImmutableList<Block> blockList = subject.getCurrentBlockList();
+
+    assertThat(blockList).isEmpty();
+  }
+
+  @Test
+  public void getCurrentBlockList_includesBlocksThatWereCompletedInThisProgram() {
+    ImmutableList<Block> blockList = subject.getCurrentBlockList();
+    assertThat(blockList).hasSize(2);
+
+    // Answer block 1 questions in this program session
+    answerNameQuestion(programDefinition.id());
+
```

Comment:
Question (blocking): should `getCurrentBlockList()` be called again here?

Reply:
Yes it should! Thank you so much for catching that!